File: platform/Flutter/sweeteditor/lib/core/core_ime.dart
// ignore_for_file: unused_element

part of 'editor_core.dart';

enum CaretAffinity {
  downstream(0),
  upstream(1);

  const CaretAffinity(this.value);
  final int value;

  static CaretAffinity fromValue(int value) {
    switch (value) {
      case 0: return downstream;
      case 1: return upstream;
      default: throw ArgumentError.value(value, 'value', 'Unknown CaretAffinity value');
    }
  }
}

enum ImeCommandKind {
  setSelection(0),
  beginComposition(1),
  updateComposition(2),
  commitText(3),
  finishComposition(4),
  cancelComposition(5),
  deleteSurrounding(6);

  const ImeCommandKind(this.value);
  final int value;

  static ImeCommandKind fromValue(int value) {
    switch (value) {
      case 0: return setSelection;
      case 1: return beginComposition;
      case 2: return updateComposition;
      case 3: return commitText;
      case 4: return finishComposition;
      case 5: return cancelComposition;
      case 6: return deleteSurrounding;
      default: throw ArgumentError.value(value, 'value', 'Unknown ImeCommandKind value');
    }
  }
}

enum ImeCoordinateSpace {
  document(0),
  editingBuffer(1),
  contextSlice(2),
  composition(3);

  const ImeCoordinateSpace(this.value);
  final int value;

  static ImeCoordinateSpace fromValue(int value) {
    switch (value) {
      case 0: return document;
      case 1: return editingBuffer;
      case 2: return contextSlice;
      case 3: return composition;
      default: throw ArgumentError.value(value, 'value', 'Unknown ImeCoordinateSpace value');
    }
  }
}

enum ImeHostAction {
  none(0),
  closeSession(1),
  restartSession(2);

  const ImeHostAction(this.value);
  final int value;

  static ImeHostAction fromValue(int value) {
    switch (value) {
      case 0: return none;
      case 1: return closeSession;
      case 2: return restartSession;
      default: throw ArgumentError.value(value, 'value', 'Unknown ImeHostAction value');
    }
  }
}

enum ImeMutationModel {
  command(0),
  textUpdate(1);

  const ImeMutationModel(this.value);
  final int value;

  static ImeMutationModel fromValue(int value) {
    switch (value) {
      case 0: return command;
      case 1: return textUpdate;
      default: throw ArgumentError.value(value, 'value', 'Unknown ImeMutationModel value');
    }
  }
}

enum ImeResultCode {
  ok(0),
  sessionMismatch(1),
  rejected(2),
  readOnly(3);

  const ImeResultCode(this.value);
  final int value;

  static ImeResultCode fromValue(int value) {
    switch (value) {
      case 0: return ok;
      case 1: return sessionMismatch;
      case 2: return rejected;
      case 3: return readOnly;
      default: throw ArgumentError.value(value, 'value', 'Unknown ImeResultCode value');
    }
  }
}

enum ImeTextSource {
  editing(0),
  committed(1),
  editingBuffer(2);

  const ImeTextSource(this.value);
  final int value;

  static ImeTextSource fromValue(int value) {
    switch (value) {
      case 0: return editing;
      case 1: return committed;
      case 2: return editingBuffer;
      default: throw ArgumentError.value(value, 'value', 'Unknown ImeTextSource value');
    }
  }
}

enum ImeTextUnit {
  utf16CodeUnit(0),
  unicodeCodePoint(1);

  const ImeTextUnit(this.value);
  final int value;

  static ImeTextUnit fromValue(int value) {
    switch (value) {
      case 0: return utf16CodeUnit;
      case 1: return unicodeCodePoint;
      default: throw ArgumentError.value(value, 'value', 'Unknown ImeTextUnit value');
    }
  }
}

class ImeCommand {
  const ImeCommand({
    this.kind = ImeCommandKind.setSelection,
    this.targetRange = const ImeOffsetRange(),
    this.selectionAfter = const ImeSelection(),
    this.text = '',
    this.deleteBefore = 0,
    this.deleteAfter = 0,
    this.textUnit = ImeTextUnit.utf16CodeUnit,
  });

  final ImeCommandKind kind;
  final ImeOffsetRange targetRange;
  final ImeSelection selectionAfter;
  final String text;
  final int deleteBefore;
  final int deleteAfter;
  final ImeTextUnit textUnit;
}

class ImeCommandBatch {
  const ImeCommandBatch({
    this.sessionId = 0,
    this.commands = const [],
  });

  final int sessionId;
  final List<ImeCommand> commands;
}

class ImeOffsetRange {
  const ImeOffsetRange({
    this.coordinateSpace = ImeCoordinateSpace.document,
    this.startUtf16 = -1,
    this.endUtf16 = -1,
  });

  final ImeCoordinateSpace coordinateSpace;
  final int startUtf16;
  final int endUtf16;
}

class ImeSelection {
  const ImeSelection({
    this.coordinateSpace = ImeCoordinateSpace.document,
    this.anchorUtf16 = -1,
    this.activeUtf16 = -1,
    this.affinity = CaretAffinity.downstream,
  });

  final ImeCoordinateSpace coordinateSpace;
  final int anchorUtf16;
  final int activeUtf16;
  final CaretAffinity affinity;
}

class ImeState {
  const ImeState({
    this.resultCode = ImeResultCode.ok,
    this.sessionId = 0,
    this.stateRevision = 0,
    this.selection = const ImeSelection(),
    this.compositionRange = const ImeOffsetRange(),
  });

  final ImeResultCode resultCode;
  final int sessionId;
  final int stateRevision;
  final ImeSelection selection;
  final ImeOffsetRange compositionRange;
}

class ImeTextContext {
  const ImeTextContext({
    this.resultCode = ImeResultCode.ok,
    this.sliceStartUtf16 = 0,
    this.totalLengthUtf16 = 0,
    this.text = '',
    this.selection = const ImeSelection(),
    this.compositionRange = const ImeOffsetRange(),
  });

  final ImeResultCode resultCode;
  final int sliceStartUtf16;
  final int totalLengthUtf16;
  final String text;
  final ImeSelection selection;
  final ImeOffsetRange compositionRange;
}

class ImeTextUpdateBatch {
  const ImeTextUpdateBatch({
    this.sessionId = 0,
    this.expectedStateRevision = 0,
    this.steps = const [],
  });

  final int sessionId;
  final int expectedStateRevision;
  final List<ImeTextUpdateStep> steps;
}

class ImeTextUpdateStep {
  const ImeTextUpdateStep({
    this.oldText = '',
    this.patchRange = const ImeOffsetRange(),
    this.replacementText = '',
    this.selectionAfter = const ImeSelection(),
    this.compositionAfter = const ImeOffsetRange(),
  });

  final String oldText;
  final ImeOffsetRange patchRange;
  final String replacementText;
  final ImeSelection selectionAfter;
  final ImeOffsetRange compositionAfter;
}

File: platform/Apple/SweetEditor-Shared/core/CoreIme.swift
import Foundation

public enum ImeCommandKind: Int32 {
    case SET_SELECTION = 0
    case BEGIN_COMPOSITION = 1
    case UPDATE_COMPOSITION = 2
    case COMMIT_TEXT = 3
    case FINISH_COMPOSITION = 4
    case CANCEL_COMPOSITION = 5
    case DELETE_SURROUNDING = 6

    public static func fromValue(_ value: Int32) -> ImeCommandKind? {
        switch value {
        case 0: return .SET_SELECTION
        case 1: return .BEGIN_COMPOSITION
        case 2: return .UPDATE_COMPOSITION
        case 3: return .COMMIT_TEXT
        case 4: return .FINISH_COMPOSITION
        case 5: return .CANCEL_COMPOSITION
        case 6: return .DELETE_SURROUNDING
        default: return nil
        }
    }
}

public enum ImeCoordinateSpace: Int32 {
    case DOCUMENT = 0
    case EDITING_BUFFER = 1
    case CONTEXT_SLICE = 2
    case COMPOSITION = 3

    public static func fromValue(_ value: Int32) -> ImeCoordinateSpace? {
        switch value {
        case 0: return .DOCUMENT
        case 1: return .EDITING_BUFFER
        case 2: return .CONTEXT_SLICE
        case 3: return .COMPOSITION
        default: return nil
        }
    }
}

public enum ImeHostAction: Int32 {
    case NONE = 0
    case CLOSE_SESSION = 1
    case RESTART_SESSION = 2

    public static func fromValue(_ value: Int32) -> ImeHostAction? {
        switch value {
        case 0: return .NONE
        case 1: return .CLOSE_SESSION
        case 2: return .RESTART_SESSION
        default: return nil
        }
    }
}

public enum ImeMutationModel: Int32 {
    case COMMAND = 0
    case TEXT_UPDATE = 1

    public static func fromValue(_ value: Int32) -> ImeMutationModel? {
        switch value {
        case 0: return .COMMAND
        case 1: return .TEXT_UPDATE
        default: return nil
        }
    }
}

public enum ImeResultCode: Int32 {
    case OK = 0
    case SESSION_MISMATCH = 1
    case REJECTED = 2
    case READ_ONLY = 3

    public static func fromValue(_ value: Int32) -> ImeResultCode? {
        switch value {
        case 0: return .OK
        case 1: return .SESSION_MISMATCH
        case 2: return .REJECTED
        case 3: return .READ_ONLY
        default: return nil
        }
    }
}

public enum ImeTextSource: Int32 {
    case EDITING = 0
    case COMMITTED = 1
    case EDITING_BUFFER = 2

    public static func fromValue(_ value: Int32) -> ImeTextSource? {
        switch value {
        case 0: return .EDITING
        case 1: return .COMMITTED
        case 2: return .EDITING_BUFFER
        default: return nil
        }
    }
}

public enum ImeTextUnit: Int32 {
    case UTF16_CODE_UNIT = 0
    case UNICODE_CODE_POINT = 1

    public static func fromValue(_ value: Int32) -> ImeTextUnit? {
        switch value {
        case 0: return .UTF16_CODE_UNIT
        case 1: return .UNICODE_CODE_POINT
        default: return nil
        }
    }
}

public struct ImeCommand {
    public var kind: ImeCommandKind = .SET_SELECTION
    public var target_range: ImeOffsetRange = ImeOffsetRange()
    public var selection_after: ImeSelection = ImeSelection()
    public var text: String = ""
    public var delete_before: Int64 = 0
    public var delete_after: Int64 = 0
    public var text_unit: ImeTextUnit = .UTF16_CODE_UNIT

    public init(kind: ImeCommandKind = .SET_SELECTION, target_range: ImeOffsetRange = ImeOffsetRange(), selection_after: ImeSelection = ImeSelection(), text: String = "", delete_before: Int64 = 0, delete_after: Int64 = 0, text_unit: ImeTextUnit = .UTF16_CODE_UNIT) {
        self.kind = kind
        self.target_range = target_range
        self.selection_after = selection_after
        self.text = text
        self.delete_before = delete_before
        self.delete_after = delete_after
        self.text_unit = text_unit
    }
}

public struct ImeCommandBatch {
    public var session_id: Int64 = 0
    public var commands: [ImeCommand] = []

    public init(session_id: Int64 = 0, commands: [ImeCommand] = []) {
        self.session_id = session_id
        self.commands = commands
    }
}

public struct ImeOffsetRange {
    public var coordinate_space: ImeCoordinateSpace = .DOCUMENT
    public var start_utf16: Int64 = -1
    public var end_utf16: Int64 = -1

    public init(coordinate_space: ImeCoordinateSpace = .DOCUMENT, start_utf16: Int64 = -1, end_utf16: Int64 = -1) {
        self.coordinate_space = coordinate_space
        self.start_utf16 = start_utf16
        self.end_utf16 = end_utf16
    }
}

public struct ImeSelection {
    public var coordinate_space: ImeCoordinateSpace = .DOCUMENT
    public var anchor_utf16: Int64 = -1
    public var active_utf16: Int64 = -1
    public var affinity: CaretAffinity = .DOWNSTREAM

    public init(coordinate_space: ImeCoordinateSpace = .DOCUMENT, anchor_utf16: Int64 = -1, active_utf16: Int64 = -1, affinity: CaretAffinity = .DOWNSTREAM) {
        self.coordinate_space = coordinate_space
        self.anchor_utf16 = anchor_utf16
        self.active_utf16 = active_utf16
        self.affinity = affinity
    }
}

public struct ImeState {
    public var result_code: ImeResultCode = .OK
    public var session_id: Int64 = 0
    public var state_revision: Int64 = 0
    public var selection: ImeSelection = ImeSelection()
    public var composition_range: ImeOffsetRange = ImeOffsetRange()

    public init(result_code: ImeResultCode = .OK, session_id: Int64 = 0, state_revision: Int64 = 0, selection: ImeSelection = ImeSelection(), composition_range: ImeOffsetRange = ImeOffsetRange()) {
        self.result_code = result_code
        self.session_id = session_id
        self.state_revision = state_revision
        self.selection = selection
        self.composition_range = composition_range
    }
}

public struct ImeTextContext {
    public var result_code: ImeResultCode = .OK
    public var slice_start_utf16: Int64 = 0
    public var total_length_utf16: Int64 = 0
    public var text: String = ""
    public var selection: ImeSelection = ImeSelection()
    public var composition_range: ImeOffsetRange = ImeOffsetRange()

    public init(result_code: ImeResultCode = .OK, slice_start_utf16: Int64 = 0, total_length_utf16: Int64 = 0, text: String = "", selection: ImeSelection = ImeSelection(), composition_range: ImeOffsetRange = ImeOffsetRange()) {
        self.result_code = result_code
        self.slice_start_utf16 = slice_start_utf16
        self.total_length_utf16 = total_length_utf16
        self.text = text
        self.selection = selection
        self.composition_range = composition_range
    }
}

public struct ImeTextUpdateBatch {
    public var session_id: Int64 = 0
    public var expected_state_revision: Int64 = 0
    public var steps: [ImeTextUpdateStep] = []

    public init(session_id: Int64 = 0, expected_state_revision: Int64 = 0, steps: [ImeTextUpdateStep] = []) {
        self.session_id = session_id
        self.expected_state_revision = expected_state_revision
        self.steps = steps
    }
}

public struct ImeTextUpdateStep {
    public var old_text: String = ""
    public var patch_range: ImeOffsetRange = ImeOffsetRange()
    public var replacement_text: String = ""
    public var selection_after: ImeSelection = ImeSelection()
    public var composition_after: ImeOffsetRange = ImeOffsetRange()

    public init(old_text: String = "", patch_range: ImeOffsetRange = ImeOffsetRange(), replacement_text: String = "", selection_after: ImeSelection = ImeSelection(), composition_after: ImeOffsetRange = ImeOffsetRange()) {
        self.old_text = old_text
        self.patch_range = patch_range
        self.replacement_text = replacement_text
        self.selection_after = selection_after
        self.composition_after = composition_after
    }
}
